src/lib/notifier.c

```diff
@@ -14,82 +14,163 @@
 #include <sof/spinlock.h>
 #include <ipc/topology.h>
 
-static struct notify_data _notify_data __aligned(PLATFORM_DCACHE_ALIGN);
-
-void notifier_register(struct notifier *notifier)
+struct notify_data {
+	void *caller;
+	enum notify_id type;
+	uint32_t data_size;
+	void *data;
+} __aligned(PLATFORM_DCACHE_ALIGN);
+
+static struct notify_data _notify_data[PLATFORM_CORE_COUNT]
+	__aligned(PLATFORM_DCACHE_ALIGN);
+
+struct callback_handle {
+	void *receiver;
+	void *caller;
+	void (*cb)(void *arg, enum notify_id, void *data);
+	struct list_item list;
+};
+
+int notifier_register(void *receiver, void *caller, enum notify_id type,
+		       void (*cb)(void *arg, enum notify_id type, void *data))
 {
 	struct notify *notify = *arch_notify_get();
+	struct callback_handle *handle;
+
+	if (type >= NOTIFIER_ID_MAX)
+		return -EINVAL;
+
+	handle = rzalloc(RZONE_SYS_RUNTIME, SOF_MEM_CAPS_RAM, sizeof(*handle));
+
+	if (!handle)
+		return -ENOMEM;
+
+	list_init(&handle->list);
+	handle->receiver = receiver;
+	handle->caller = caller;
+	handle->cb = cb;
 
 	spin_lock(notify->lock);
-	list_item_prepend(&notifier->list, &notify->list);
+	list_item_prepend(&handle->list, &notify->list[type]);
 	spin_unlock(notify->lock);
+
+	return 0;
 }
 
-void notifier_unregister(struct notifier *notifier)
+void notifier_unregister(void *receiver, void *caller, enum notify_id type)
 {
 	struct notify *notify = *arch_notify_get();
-
+	struct list_item *wlist;
+	struct list_item *tlist;
+	struct callback_handle *handle;
+
+	if (type >= NOTIFIER_ID_MAX)
+		return;
+
+	/*
+	 * Unregister all matching callbacks
+	 * If receiver is NULL, unregister all callbacks with matching callers
+	 * If caller is NULL, unregister all callbacks with matching receivers
+	 *
+	 * Event producer might force unregister all receivers by passing
+	 * receiver NULL
+	 * Event consumer might unregister from all callers by passing caller
+	 * NULL
+	 */
 	spin_lock(notify->lock);
-	list_item_del(&notifier->list);
+	list_for_item_safe(wlist, tlist, &notify->list[type]) {
+		handle = container_of(wlist, struct callback_handle, list);
+		if ((!receiver || handle->receiver == receiver) &&
+		    (!caller || handle->caller == caller)) {
+			list_item_del(&handle->list);
+			rfree(handle);
+		}
+	}
 	spin_unlock(notify->lock);
 }
 
-void notifier_notify(void)
+void notifier_unregister_all(void *receiver, void *caller)
+{
+	int i;
+
+	for (i = 0; i < NOTIFIER_ID_MAX; i++)
```

Comment:
Why do not init with the first symbol from enum list?

Reply:
done